Fix #8189 - refactor search.c

MaskRay opened the pull request:
* fix r_search_mybinparse_update when locating `a+a+b` in `a+a+a+b` where len(a) > 1, e.g. locating `ababc` in `abababc`
* swap the order of the nested loop from

  for (i = 0; i < len; i++)
    r_list_foreach (s->kws, iter, kw)

  to

  r_list_foreach (s->kws, iter, kw)
    for (i = 0; i < len; i++)

  rationale:
  + searching for more needles is rare. the branches used in inner loop harms performance
  + it's cumbersome to bookkeep for each kw in advanced searching algorithms



Fixed tests:

```
[FX]  42  cmd_search: /x with bin mask
[FX]  43  cmd_search: /x with bin mask
```

radare commented:
Yeah 3 fixes... but 23 XX

> On 12 Aug 2017, at 07:25, Fangrui Song <notifications@github.com> wrote:
> 
> fix r_search_mybinparse_update when locating a+a+b in a+a+a+b where len(a) > 1, e.g. locating ababc in abababc
> 
> remove questionable config var search.distance because it returns
> false positive almost everywhere when search.distance > 0
> 
> swap the order of the nested loop from
> 
> r_list_foreach (s->kws, iter, kw)
> for (i = 0; i < len; i++)
> 
> to
> 
> for (i = 0; i < len; i++)
> r_list_foreach (s->kws, iter, kw)
> 
> rationale:
> 
> searching for more needles is rare. the branches used in inner loop harms performance
> it's cumbersome to bookkeep for each kw in advanced searching algorithms
> r_search_mybinparse_update crossing boundaries is hard and not treated
> in this commit.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/radare/radare2/pull/8195
> 
> Commit Summary
> 
> Fix #8189 - refactor search.c
> File Changes
> 
> M libr/core/cmd_search.c (23)
> M libr/include/r_search.h (6)
> M libr/search/search.c (287)
> Patch Links:
> 
> https://github.com/radare/radare2/pull/8195.patch
> https://github.com/radare/radare2/pull/8195.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

MaskRay commented:
Reverted this part:

```
diff --git c/libr/core/cmd_search.c i/libr/core/cmd_search.c                                                                                                              
index b8adf2e66..48ea3241d 100644                                                                                                                                         
--- c/libr/core/cmd_search.c                                                                                                                                              
+++ i/libr/core/cmd_search.c                                                                                                                                              
@@ -2010,7 +2010,10 @@ static void do_string_search(RCore *core, struct search_parameters *param) {                                                                       
                                if ((at + bufsz) > param->to) {                                                                                                           
                                        bufsz = param->to - at;                                                                                                           
                                }         
-                               ret = r_io_read_at (core->io, at, buf, bufsz);       
+                               // XXX cause  29  cmd_search: /x wrong entries to fail                                                                                    
+                               // if seek fails we shouldnt read at all             
+                               (void) r_io_seek (core->io, at, R_IO_SEEK_SET);      
+                               ret = r_io_read (core->io, buf, bufsz);  
```

Should be good now but with one less fix.

radare commented:
Still red

XVilka commented on libr/search/search.c:
```diff
@@ -427,7 +348,7 @@ R_API RList *r_search_find(RSearch *s, ut64 addr, const ut8 *buf, int len) {
 /* --- keywords --- */
 R_API int r_search_kw_add(RSearch *s, RSearchKeyword *kw) {
-	if (!kw) return false;
+	if (!kw->keyword_length) return false;
```
check both, `kw` can be NULL in theory. Same thing for `s`

MaskRay replied:
Unnecessary as call sites have checked this

radare replied:
missing braces

MaskRay replied:
IMHO using curly braces everywhere does not enhance readability, especially for short conditional statements such as:
```c
if (!kw->keyword_length) {
  return false;
}
```

Linux coding style allows:
```c
        if (!buffer)
                return -ENOMEM;
```

Google C++ Style Guide (https://google.github.io/styleguide/cppguide.html#Conditionals) allows the statement put on the same line if it enhances readability:

```c
if (x == kFoo) return new Foo();
```

radare replied:
Not the place to discuss this and it was already discussed several times. You are late to the party and we spent 2 years cooking a coding style that fits well for most of us. So just follow it

XVilka commented on libr/search/search.c:
```diff
@@ -439,11 +360,11 @@ R_API void r_search_kw_reset(RSearch *s) {
 }
 
 R_API void r_search_reset(RSearch *s, int mode) {
+	R_FREE (s->data);
```
Check `s` for NULL before dereferencing.

MaskRay replied:
This usages is like calling a member function where the first argument denotes the instance. The call site should make this check and it would be too defensive to add such checks.

XVilka replied:
This is not C++.

MaskRay replied:
The check was not there before my change. Do that in another commit if you really care.

radare replied:
Add the check

radare commented on libr/core/cmd_search.c:
```diff
@@ -1924,6 +1924,9 @@ static void do_string_search(RCore *core, struct search_parameters *param) {
 	}
 	core->search->inverse = param->inverse;
+	if (param->inverse) {
+		core->search->maxhits = 1;
```
Why inverse search forces maxhits to 1?

MaskRay replied:
This is exactly what the code did before.

MaskRay replied:
This behavior was implemented in `R_API int r_search_mybinparse_update(void *_s, ut64 from, const ut8 *buf, int len) {` and there are tests testing this. I just moved it outside .

radare replied:
Just keep the old behaviour. Theres no sense in searching for 0 keywords and ut64max is such a bad number to type for a human to specify that theres no limit

MaskRay replied:
The old behavior is stopping at the first occurrence:

```c
-                       /* Stop at the first occurrence */
-                       if (s->inverse) {
-                               return -1;
-                       }
```

Then `r_search_update` returns `-1` and `do_string_search` in cmd_search.c breaks the loop.

radare commented on libr/core/cmd_search.c:
```diff
@@ -2349,7 +2353,11 @@ static int cmd_search(void *data, const char *input) {
-	maxhits = r_config_get_i (core->config, "search.maxhits");
+	core->search->maxhits = r_config_get_i (core->config, "search.maxhits");
+	// TODO default search.maxhits to UT64_MAX
+	if (!core->search->maxhits) {
```
Wrong. Maxhits=0 is no limit, ut64max is a big limit

MaskRay replied:
`ut64max` can be seen as infinity and the limit cannot be reached. This simplifies code otherwhere and there would be many more other issues before the limit was reached.

radare replied:
imho having 0 maxhits is more logic and easy to initialize than a value that is not zero. and it was already like this. also by the user it is easier to type e search.maxhits=0 than e search.maxhits=0xffffffffffffffffff

MaskRay replied:
If `search.maxhits` is zero (default), `core->search->maxhits` will be `UT64MAX`.

radare commented on libr/include/r_search.h:
```diff
@@ -34,8 +34,6 @@ typedef struct r_search_keyword_t {
 	ut8 *bin_binmask;
 	ut32 keyword_length;
 	ut32 binmask_length;
-	ut32 idx[R_SEARCH_DISTANCE_MAX]; // searching purposes
```
Why do u remove a feature instead of fixing it?

MaskRay replied:
It was implemented in such a wrong way that I could not guess its intention.

MaskRay replied:
Brought back.

radare replied:
we can discuss this thing in a separate issue/pr, better change only 1 thing on each PR, this way it's easier to review

MaskRay replied:
`idx[R_SEARCH_DISTANCE_MAX]` gets in the way when I made the PR. It introduced another nested loop which I did not know how to put it back in the code after the refactoring as the loop order had been swapped.

radare commented on libr/search/search.c:
```diff
@@ -125,258 +131,172 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
+	r_list_foreach (s->kws, iter, kw) {
+		ut8 *a = kw->bin_keyword;
+		for (i = 0; i + kw->keyword_length < len; i++) {
+			if ((ut8)(buf[i+1] - buf[i]) == a[0]) {
```
Strange arithmetics

MaskRay replied:
`/d` was incorrect if `/d 0101` was intended to find `01 02 03`. This is corrected now.

radare replied:
can you add tests for this or unmark the BROKEN in r2r? ill merge that before merging this.

thanks

MaskRay replied:
https://github.com/radare/radare2-regressions/pull/959

I do not see how to make `/d` output one more byte for each hit.

radare commented on libr/search/search.c:
```diff
@@ -125,258 +131,172 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
+		ut8 *a = kw->bin_keyword;
+		for (i = 0; i + kw->keyword_length < len; i++) {
+			if ((ut8)(buf[i+1] - buf[i]) == a[0]) {
+				for (j = 1; j < kw->keyword_length &&
```
Ugly for

MaskRay replied:
Move `j++` inside of `{}`?

MaskRay replied:
Fixed but not yet pushed to wait for travis.

radare commented on libr/search/search.c:
```diff
@@ -125,258 +131,172 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
-	if (kw_len < 0)
-		return false;
-
-	for (i = 0; i < 256; i++)
```
Missing braces

MaskRay replied:
Fixed but not yet pushed to wait for travis.

MaskRay replied:
Pushed

radare commented on libr/search/search.c:
```diff
@@ -125,258 +131,172 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
-		pos += bad_char_shift[kw->icase?tolower(ch):ch];
+		ch = buf[i + m];
+		i += bad_char_shift[kw->icase ? tolower (ch) : ch];
+	next:;
```
Labels doesnt require ;

MaskRay replied:
The last statement in a scope requires `;`.

radare replied:
it will be probably easier to read if you put this label at the begining of the loop. also we may assume i>0 otherwise this can result in an infinite loop.

MaskRay replied:
Moved `next` to the beginning of the loop.

`i>=0` is a loop invariant.

radare commented on libr/search/search.c:
```diff
@@ -42,6 +50,7 @@ R_API RSearch *r_search_new(int mode) {
 R_API RSearch *r_search_free(RSearch *s) {
 	if (!s) return NULL;
```
missing braces

MaskRay replied:
IMHO the following two alternatives enhance readability. Also see my comment above.
```c
if (!s) return NULL;
```

```c
if (!s)
  return NULL;
```

radare replied:
See coding style and be consistent. This has been discussed zillions of times and im not gonna explain the reasons again

radare commented on libr/search/search.c:
```diff
@@ -125,250 +132,203 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
-			}
+		longest = R_MAX (longest, kw->keyword_length);
+	}
+	if (!longest) return 0;
```
missing braces

radare commented on libr/search/search.c:
```diff
@@ -125,250 +132,203 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
 		}
+	} else {
+		left = malloc (sizeof(RSearchLeftover) + 2 * (longest - 1));
+		if (!left) return -1;
```
missing braces

radare commented on libr/search/search.c:
```diff
@@ -125,250 +132,203 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
+				if (!r_search_hit_new (s, kw, from + i)) return -1;
+				kw->count++;
+				count++;
+				if (++s->nhits >= s->maxhits) return count;
```
missing braces

radare commented on libr/search/search.c:
```diff
@@ -125,250 +132,203 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
-					}
+		for (i = 0; i + kw->keyword_length <= len; i++) {
+			if (brute_force_match(s, kw, buf, i) != s->inverse) {
+				if (!r_search_hit_new (s, kw, from + i)) return -1;
```
missing braces

radare commented on libr/search/search.c:
```diff
@@ -125,250 +132,203 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
-						//s->nhits++;
-					}
+		for (i = 0; i + kw->keyword_length <= len; i++) {
+			if (brute_force_match(s, kw, buf, i) != s->inverse) {
```
missing space before (

MaskRay replied:
Fixed to be consistent with other code, though I always think space between function name and `(args` is bizarre...

MaskRay commented:
Resolved for most style issues.

`if (!s) return NULL;`

Though mandatory curly braces around simple conditional statements are still unacceptable to me.
```c
if (!s) {
  return NULL;
}
`

MaskRay commented:
Rebased on HEAD and added braces.

radare commented:
The braces in oneliner blocks they look bad indeed. But they clearly show something that is important and eases the reading, also exposes something that could be expressed in a better way by following functional programming or raii techniques. Also gotos are not considered a good practicw becayse makes the reader jump up and down randomly. But they sometimes help on reducing code that involves error checking.

Ive been coding for more than 20 years and those are some insights i've learned. I would never do those things 15 years ago. So consider it an advice for the future and for consistency. Perl for example imposes braces on all blocks. Even one liners and the reasons vary, but one of them is readability.

Your future yourself will be pleased to follow some consistent coding styles. Trust me ;)

radare commented on libr/search/search.c:
```diff
@@ -125,250 +134,217 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
-	if (kw->icase) {
-		a = tolower (a);
-		b = tolower (b);
+static inline bool brute_force_match(RSearch *s, RSearchKeyword *kw, const ut8 *buf, int i) {
```
Inlining a 100 line function? No thanks

MaskRay replied:
Done

radare commented on libr/core/cmd_search.c:
```diff
@@ -2352,7 +2356,10 @@ static int cmd_search(void *data, const char *input) {
  */
-	maxhits = r_config_get_i (core->config, "search.maxhits");
+	core->search->maxhits = r_config_get_i (core->config, "search.maxhits");
+	if (!core->search->maxhits) {
```
Remove this

MaskRay replied:
Removed

radare commented on libr/core/cmd_search.c:
```diff
@@ -2716,7 +2723,7 @@ static int cmd_search(void *data, const char *input) {
 			json = true;
 		}
-	/* pass-thru */
+		// fallthrough
```
The reason for that comment is to avoid an error in scan.coverity not sure if that change will be also accepted by them but we can try

MaskRay replied:
https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html -Wimplicit-fallthrough=n

`fallthrough` is recognized but not `pass-thru`.

radare replied:
oki :)

radare commented on libr/search/search.c:
```diff
@@ -28,6 +35,7 @@ R_API RSearch *r_search_new(int mode) {
 	s->string_max = 255;
 	s->string_min = 3;
 	s->hits = r_list_new ();
+	s->maxhits = UT64_MAX;
```
Should be =0

radare commented on libr/search/search.c:
```diff
@@ -125,250 +134,217 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
+			left->len = 0;
+		}
+	} else {
+		left = malloc (sizeof(RSearchLeftover) + 2 * (longest - 1));
```
Missing space and maybe use calloc instead to avoid uninit buffer

radare replied:
Longest can be <2? If so that code can bring some problems

radare replied:
Better use calloc instead or * to avoid integer overflows

MaskRay replied:
Its `sizeof(RSearchLeftover) + 2 * (longest - 1)`, more disasters should have happened before the integer overflow kicks in.

radare replied:
its not about the worst case scenario, its about good practices

MaskRay replied:
Changed to `		left = malloc (sizeof(RSearchLeftover) + (size_t)2 * (longest - 1));`

It will never cause trouble where `size_t` is 64-bit and `longest` is 32-bit.

radare commented on libr/search/search.c:
```diff
@@ -125,250 +134,217 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
+				}
+				kw->count++;
+				count++;
+				if (++s->nhits >= s->maxhits) {
```
Im not a big fan of preincrements inside if conditionals better put yhat in the line above

MaskRay replied:
Done

radare commented on libr/search/search.c:
```diff
@@ -427,7 +406,9 @@ R_API RList *r_search_find(RSearch *s, ut64 addr, const ut8 *buf, int len) {
 
 /* --- keywords --- */
 R_API int r_search_kw_add(RSearch *s, RSearchKeyword *kw) {
-	if (!kw) return false;
```
Null deref regression

radare replied:
Data should never be null.

radare commented on libr/search/search.c:
```diff
@@ -439,11 +420,11 @@ R_API void r_search_kw_reset(RSearch *s) {
-	s->nhits = 0;
-	s->hits = r_list_new ();
+	s->hits = r_list_newf (free);
 	if (!s->hits) return;
```
Missing braces

MaskRay replied:
Done ( though i hadn't touched it)

radare commented on libr/search/search.c:
```diff
@@ -439,11 +420,11 @@ R_API void r_search_kw_reset(RSearch *s) {
-	s->hits->free = free;
+	s->nhits = 0;
 	r_search_kw_reset (s);
 	if (!r_search_set_mode (s, mode))
```
Missing braces

MaskRay replied:
Done

MaskRay commented on libr/search/search.c:
```diff
@@ -427,7 +406,9 @@ R_API RList *r_search_find(RSearch *s, ut64 addr, const ut8 *buf, int len) {
 
 /* --- keywords --- */
 R_API int r_search_kw_add(RSearch *s, RSearchKeyword *kw) {
-	if (!kw) return false;
```
Now `	if (!kw || !kw->keyword_length) {`

I believe the call site should do the check, not here.

radare replied:
an if doesnst hurts and prevents anoying null derefs, compilers can always optimize that stuff with lto

MaskRay replied:
The call sites should do the check.

```c
	case 'a': if (input[1]) {
...........
				r_search_kw_add (core->search,
					r_search_keyword_new_hexmask (kwd, NULL)); ///// it should check if new_hexmask returns NULL
				r_search_begin (core->search);
				free (kwd);
				dosearch = true;
```

The call site are closer to the user and have more knowledge about the interactive context. It can give useful warnings/errors and it should have do that. Then such check in the search.c side will be redundant. If the library is full of such `if (!xxx) return;` checks, there will not be much integrity guarantee and you'll tend to distrust everything.

MaskRay replied:
I was really, really sad when I saw the following code for the first time:

```c
#define r_list_foreach(list, it, pos)\
	if (list)\
		for (it = list->head; it && (pos = it->data); it = it->n)
```

`pos = it->data` is dangerous, what if one of the elements is `NULL`? It will stop iterating at that element, ignoring the rest.

`if (list)` will you miss something if the for loop has not been executed? The (no-null-dereference) illusion comes with the cost that more trouble may accompany.

radare commented on libr/search/search.c:
```diff
@@ -9,6 +9,12 @@
+typedef struct {
+	ut64 end;
+	int len;
+	ut8 data[];
```
use ut8 *data

MaskRay replied:
`ut8 *data` (avoiding flexible array member) leads to one extra `malloc`, more error check with no benefit.

MaskRay replied:
It is also supported by tcc.

condret replied:
use * instead of [] , it is more portable, especially for older compilers. Use [] only for fixed sizes

MaskRay replied:
I'd like to accept a feature if compiles on gcc, clang and MSVC 2015.
Without  flexible array member, you'll do two `malloc`, the second for `ut8 *data`, and if it returned `NULL`, you had to clean the struct and return `NULL`.

MaskRay replied:
Yes, I'm also talking to `r_strbuf`

```c
R_API bool r_strbuf_set(RStrBuf *sb, const char *s) {
	int l;
	if (!sb) {
		return false;
	}
	if (!s) {
		r_strbuf_init (sb);
		return true;
	}
	l = strlen (s);
	if (l >= sizeof (sb->buf)) {
		char *ptr = sb->ptr;
		if (!ptr || l+1 > sb->ptrlen) {
			ptr = malloc (l + 1);
			if (!ptr) {
				return false;
			}
			sb->ptrlen = l + 1;
			sb->ptr = ptr;
		}
		memcpy (ptr, s, l+1);
	} else {
		sb->ptr = NULL;
		memcpy (sb->buf, s, l+1);
	}
	sb->len = l;
	return true;
}
```

MaskRay commented:
```c
typedef struct {
	int len;
	char *ptr;
	int ptrlen;
	char buf[64];
} RStrBuf;

R_API bool r_strbuf_set(RStrBuf *sb, const char *s) {
	int l;
	if (!sb) {
		return false;
	}
	if (!s) {
		r_strbuf_init (sb);
		return true;
	}
	l = strlen (s);
	if (l >= sizeof (sb->buf)) {
		char *ptr = sb->ptr;
		if (!ptr || l+1 > sb->ptrlen) {
			ptr = malloc (l + 1);
			if (!ptr) {
				return false;
			}
			sb->ptrlen = l + 1;
			sb->ptr = ptr;
		}
		memcpy (ptr, s, l+1);
	} else {
		sb->ptr = NULL;
		memcpy (sb->buf, s, l+1);
	}
	sb->len = l;
	return true;
}
```

This can be avoided if flexible array member is used.

radare commented:
you cant resize a flexible array member, you have to resize the whole pointer, so ends up breaking the api because you would need to pass RStrBuf** or return the new RStrBufer delaying the responsability to control the new pointer to the caller. Seems like msvc is also supporting this, and this will reduce the memory usage, but breaking api is probably the main concern for me

> On 17 Aug 2017, at 21:54, Fangrui Song <notifications@github.com> wrote:
> 
> typedef struct {
> 	int len;
> 	char *ptr;
> 	int ptrlen;
> 	char buf[64];
> } RStrBuf;
> 
> R_API bool r_strbuf_set(RStrBuf *sb, const char *s) {
> 	int l;
> 	if (!sb) {
> 		return false;
> 	}
> 	if (!s) {
> 		r_strbuf_init (sb);
> 		return true;
> 	}
> 	l = strlen (s);
> 	if (l >= sizeof (sb->buf)) {
> 		char *ptr = sb->ptr;
> 		if (!ptr || l+1 > sb->ptrlen) {
> 			ptr = malloc (l + 1);
> 			if (!ptr) {
> 				return false;
> 			}
> 			sb->ptrlen = l + 1;
> 			sb->ptr = ptr;
> 		}
> 		memcpy (ptr, s, l+1);
> 	} else {
> 		sb->ptr = NULL;
> 		memcpy (sb->buf, s, l+1);
> 	}
> 	sb->len = l;
> 	return true;
> }
> This can be avoided if flexible array member is used.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/radare/radare2/pull/8195#issuecomment-323176859>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AA3-lnx8I_MUL_J-0wj25Iw_VK4FKzD4ks5sZJp3gaJpZM4O1URX>.
>

radare commented on libr/search/search.c:
```diff
@@ -127,250 +136,220 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
 	RSearch *s = (RSearch*)_s;
+	RSearchKeyword *kw;
 
-	for (i=0; i<len; i++) {
```
so you are only checking  the first char of the buffer, not all of them?

MaskRay replied:
The code does check all of them.

```c
		for (i = 0; i + kw->keyword_length < len; i++) {
			if ((ut8)(buf[i+1] - buf[i]) == a[0]) {
				j = 1;
				while (j < kw->keyword_length && (ut8)(buf[i+j+1] - buf[i+j]) == a[j]) {
					j++;
				}
				if (j == kw->keyword_length) { ////////////// all of them

```

radare commented:
once master gets green again i would like this pr to be rebased and i'l have a look again, sorry for the lag, crazy days irl

MaskRay commented:
It makes things simpler if this PR can be decided "to be or not to be" earlier. There are also new challenges when siol lands. Stacking maps need different treatment other than the current approach.

MaskRay commented:
Rebased on master.

@alvarofe Your recent `bool r_io_read_at` stuff makes search more broken.

condret commented:
@MaskRay use the al_read here, this is the way to go

MaskRay commented:
Rebased.

A
G
A
I
N

radare commented on libr/core/cmd_search.c:
```diff
@@ -1854,6 +1854,10 @@ static void do_string_search(RCore *core, struct search_parameters *param) {
 	core->search->inverse = param->inverse;
+	// TODO Bad but is to be compatible with the legacy behavior
+	if (param->inverse) {
+		core->search->maxhits = 1;
```
Why

radare commented:
im merging, but im not happy with removing features or hardcoding maxhits, but this thing got too far. please FOCUS, we need to fix all the issues related to the siol merge, and without having travis green its a pain to keep merging more stuff that cant be properly checked without that. several more tests needs to be updated after my changes in search.in to make it correct.

MaskRay commented:
I was not removing features or harcoding maxhits (it existed before the refactor)

`maxhits` was used as a decreasing counter. I made it constant in the search loop and added another field `nhits` to record the number of hits.
I was not removing features. The so-called "search purpose" was used to implement `search.distance`, but it was incorrect and harmed the performance drastically (O(|haystack|*|needle|*|search.distance|), with some coined test cases, it fell into an infinite loop). So I had to remove it and replaced it with a seemingly-correct leftover.

radare commented:
Ok

> On 13 Sep 2017, at 18:00, Fangrui Song <notifications@github.com> wrote:
> 
> I was not removing features or harcoding maxhits (it existed before the refactor)
> 
> maxhits was used as a decreasing counter. I made it constant in the search loop and added another field nhits to record the number of hits.
> I was not removing features. The so-called "search purpose" was used to implement search.distance, but it was incorrect and harmed the performance drastically (O(|haystack||needle||search.distance|), with some coined test cases, it fell into an infinite loop). So I had to remove it and replaced it with a seemingly-correct leftover.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
